File: client/lib/services/clerk_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../config/clerk_config.dart';

/// Clerk Authentication Service
///
/// Handles all Clerk authentication operations including:
/// - Sign in with email/password
/// - Sign up with email/password
/// - Two-factor authentication (2FA) via email PIN
/// - Email verification
/// - Session management
/// - Sign out
class ClerkService {
  static final ClerkService _instance = ClerkService._internal();
  factory ClerkService() => _instance;
  ClerkService._internal();

  // Note: Clerk Flutter SDK does not have a stable API yet
  // We'll use HTTP requests directly to Clerk's API endpoints
  bool _initialized = false;

  /// Initialize service
  Future<void> initialize() async {
    if (_initialized) return;
    _initialized = true;
  }

  /// Sign in with email and password
  ///
  /// Returns a Map with:
  /// - 'success': bool
  /// - 'requires2FA': bool (if true, call verify2FACode next)
  /// - 'sessionToken': String? (if success and no 2FA required)
  /// - 'signInId': String? (for 2FA flow)
  /// - 'error': String? (if failed)
  Future<Map<String, dynamic>> signInWithEmailPassword(String email, String password) async {
    try {
      await initialize();

      // Call Clerk API to create sign-in
      final response = await http.post(
        Uri.parse('${ClerkConfig.clerkFrontendApi}/v1/client/sign_ins'),
        headers: {
          'Content-Type': 'application/json',
        },
        body: json.encode({
          'identifier': email,
          'password': password,
        }),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final signInData = data['response'] ?? data;

        // Check if 2FA is required
        if (signInData['status'] == 'needs_second_factor') {
          // Prepare email code for 2FA
          final prepareResponse = await http.post(
            Uri.parse('${ClerkConfig.clerkFrontendApi}/v1/client/sign_ins/${signInData['id']}/prepare_second_factor'),
            headers: {'Content-Type': 'application/json'},
            body: json.encode({'strategy': 'email_code'}),
          );

          if (prepareResponse.statusCode == 200) {
            return {
              'success': false,
              'requires2FA': true,
              'signInId': signInData['id'],
              'message': '2FA code sent to your email',
            };
          }
        }

        // Check if sign-in is complete
        if (signInData['status'] == 'complete') {
          final sessionToken = signInData['last_active_session_id'];

          if (sessionToken == null) {
            return {'success': false, 'error': 'Failed to get session token'};
          }

          await _saveSessionToken(sessionToken);
          await _fetchAndSaveUserData(sessionToken);

          return {
            'success': true,
            'requires2FA': false,
            'sessionToken': sessionToken,
          };
        }

        return {'success': false, 'error': 'Sign-in incomplete'};
      }

      return {'success': false, 'error': 'Invalid credentials'};
    } catch (e) {
      if (kDebugMode) print('Sign in error: $e');
      return {'success': false, 'error': e.toString()};
    }
  }

  /// Verify 2FA code
  ///
  /// Call this after signInWithEmailPassword returns requires2FA: true
  Future<Map<String, dynamic>> verify2FACode(String signInId, String code) async {
    try {
      final response = await http.post(
        Uri.parse('${ClerkConfig.clerkFrontendApi}/v1/client/sign_ins/$signInId/attempt_second_factor'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'strategy': 'email_code',
          'code': code,
        }),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final signInData = data['response'] ?? data;

        if (signInData['status'] == 'complete') {
          final sessionToken = signInData['last_active_session_id'];

          if (sessionToken == null) {
            return {'success': false, 'error': 'Failed to get session token'};
          }

          await _saveSessionToken(sessionToken);
          await _fetchAndSaveUserData(sessionToken);

          return {
            'success': true,
            'sessionToken': sessionToken,
          };
        }
      }

      return {'success': false, 'error': 'Invalid verification code'};
    } catch (e) {
      if (kDebugMode) print('2FA verification error: $e');
      return {'success': false, 'error': e.toString()};
    }
  }

  /// Sign up with email and password
  ///
  /// Returns a Map with:
  /// - 'success': bool
  /// - 'userId': String? (Clerk user ID)
  /// - 'error': String? (if failed)
  Future<Map<String, dynamic>> signUpWithEmailPassword({
    required String email,
    required String password,
    required String firstName,
    required String lastName,
  }) async {
    try {
      await initialize();

      // Call backend endpoint to create user in Clerk
      final response = await http.post(
        Uri.parse('${ClerkConfig.backendUrl}/auth/clerk-signup'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'email': email,
          'password': password,
          'firstName': firstName,
          'lastName': lastName,
        }),
      );

      if (kDebugMode) {
        print('Sign up response status: ${response.statusCode}');
        print('Sign up response body: ${response.body}');
      }

      if (response.statusCode == 201) {
        final data = json.decode(response.body);

        return {
          'success': true,
          'userId': data['userId'],
          'message': data['message'],
        };
      } else {
        final error = json.decode(response.body);
        return {
          'success': false,
          'error': error['message'] ?? 'Sign up failed',
        };
      }
    } catch (e) {
      if (kDebugMode) print('Sign up error: $e');
      return {'success': false, 'error': 'Network error: ${e.toString()}'};
    }
  }

  /// Verify email with code
  ///
  /// Call this after signUpWithEmailPassword
  Future<Map<String, dynamic>> verifyEmail(String signUpId, String code) async {
    try {
      final response = await http.post(
        Uri.parse('${ClerkConfig.clerkFrontendApi}/v1/client/sign_ups/$signUpId/attempt_verification'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'strategy': 'email_code',
          'code': code,
        }),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final signUpData = data['response'] ?? data;

        if (signUpData['status'] == 'complete') {
          final sessionToken = signUpData['created_session_id'];

          if (sessionToken == null) {
            return {'success': false, 'error': 'Failed to get session token'};
          }

          await _saveSessionToken(sessionToken);
          await _fetchAndSaveUserData(sessionToken);

          return {
            'success': true,
            'sessionToken': sessionToken,
          };
        }
      }

      return {'success': false, 'error': 'Invalid verification code'};
    } catch (e) {
      if (kDebugMode) print('Email verification error: $e');
      return {'success': false, 'error': e.toString()};
    }
  }

  /// Sign out
  ///
  /// Revokes session with Clerk and clears local storage
  Future<bool> signOut() async {
    try {
      await initialize();

      final sessionToken = await getSessionToken();

      // Call backend to revoke session
      if (sessionToken != null) {
        try {
          await http.post(
            Uri.parse('${ClerkConfig.backendUrl}${ClerkConfig.clerkLogoutEndpoint}'),
            headers: {
              'Authorization': 'Bearer $sessionToken',
              'Content-Type': 'application/json',
            },
          );
        } catch (e) {
          if (kDebugMode) print('Failed to revoke session on backend: $e');
        }
      }

      // Clear local storage
      await _clearSession();

      return true;
    } catch (e) {
      if (kDebugMode) print('Sign out error: $e');
      return false;
    }
  }

  /// Get current session token from local storage
  Future<String?> getSessionToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(ClerkConfig.sessionTokenKey);
  }

  /// Check if user is authenticated
  Future<bool> isAuthenticated() async {
    final token = await getSessionToken();
    if (token == null) return false;

    // Optionally verify token with backend
    try {
      final response = await http.get(
        Uri.parse('${ClerkConfig.backendUrl}${ClerkConfig.authMeEndpoint}'),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }

  /// Get current user data from local storage
  Future<Map<String, dynamic>?> getUserData() async {
    final prefs = await SharedPreferences.getInstance();
    final userDataJson = prefs.getString(ClerkConfig.userDataKey);

    if (userDataJson == null) return null;

    return json.decode(userDataJson) as Map<String, dynamic>;
  }

  // ===================
  // PRIVATE HELPERS
  // ===================

  /// Save session token to local storage
  Future<void> _saveSessionToken(String token) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(ClerkConfig.sessionTokenKey, token);
  }

  /// Fetch user data from backend and save to local storage
  Future<void> _fetchAndSaveUserData(String sessionToken) async {
    try {
      final response = await http.get(
        Uri.parse('${ClerkConfig.backendUrl}${ClerkConfig.authMeEndpoint}'),
        headers: {
          'Authorization': 'Bearer $sessionToken',
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final userData = data['user'];

        // Save to local storage
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString(ClerkConfig.userDataKey, json.encode(userData));
      }
    } catch (e) {
      if (kDebugMode) print('Failed to fetch user data: $e');
    }
  }

  /// Clear session data from local storage
  Future<void> _clearSession() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(ClerkConfig.sessionTokenKey);
    await prefs.remove(ClerkConfig.userDataKey);
  }
}

File: server/services/clerkService.js
// ============================================
// DEPRECATED - DO NOT USE
// ============================================
// This file is deprecated as of the latest refactor.
// Authentication is now handled directly from the Flutter client
// calling Clerk's Frontend API, not proxied through this backend.
//
// The backend only validates Clerk session tokens via:
// - GET /auth/me (validate session and get user info)
// - POST /auth/clerk-logout (revoke session)
//
// This file is kept for reference but should not be used.
// ============================================

require("dotenv").config();
const axios = require("axios");
const { CookieJar } = require("tough-cookie");
const { wrapper } = require("axios-cookiejar-support");

// --------------------
// Constants
// --------------------
const CLERK_FRONTEND_API = process.env.CLERK_FRONTEND_API; // Get from .env

// --------------------
// Logging
// --------------------
const winston = require("winston");
const logger = winston.createLogger({
  level: "info",
  format: winston.format.combine(winston.format.timestamp(), winston.format.json()),
  transports: [
    new winston.transports.Console(),
    new winston.transports.File({ filename: "logs/clerk-service.log" }),
  ],
});

// --------------------
// Helper to create Clerk API client with cookie support
// --------------------
function createClerkApiClient() {
  const jar = new CookieJar();
  const client = wrapper(axios.create({
    baseURL: CLERK_FRONTEND_API,
    headers: {
      "Content-Type": "application/json",
    },
    jar,
    withCredentials: true,
  }));
  return client;
}

// --------------------
// Functions
// --------------------

async function signUpWithEmailPassword(email, password, firstName, lastName) {
  const client = createClerkApiClient();

  try {
    logger.info(`Starting sign-up flow for: ${email}`);

    // Step 1: Create a sign-up session using frontend API
    const signUpResponse = await client.post(
      `/v1/client/sign_ups`,
      {
        email_address: email,
        password: password,
        first_name: firstName,
        last_name: lastName,
      }
    );

    const signUpData = signUpResponse.data.response || signUpResponse.data;
    logger.info("Sign-up created:", { status: signUpData.status, id: signUpData.id });

    // Step 2: Prepare email verification
    await client.post(
      `/v1/client/sign_ups/${signUpData.id}/prepare_verification`,
      { strategy: "email_code" }
    );

    logger.info("Email verification prepared for:", email);

    return {
      success: false,
      requiresVerification: true,
      signUpId: signUpData.id,
      message: "Verification code sent to your email",
    };
  } catch (error) {
    logger.error("Error in signUpWithEmailPassword:", error);
    const errorMessage = error.response?.data?.errors?.[0]?.message || error.message;
    return { success: false, error: errorMessage || "Failed to create user." };
  }
}

async function signInWithEmailPassword(email, password) {
  const client = createClerkApiClient();

  try {
    logger.info(`Attempting sign-in with Clerk frontend API: ${email}`);

    // Step 1: Create sign-in with identifier
    const createResponse = await client.post(
      `/v1/client/sign_ins`,
      { identifier: email }
    );

    const signInData = createResponse.data.response || createResponse.data;
    logger.info("Sign-in created:", { status: signInData.status, id: signInData.id });

    // Step 2: Attempt password authentication
    const attemptResponse = await client.post(
      `/v1/client/sign_ins/${signInData.id}/attempt_first_factor`,
      { strategy: "password", password }
    );

    const attemptData = attemptResponse.data.response || attemptResponse.data;
    logger.info("Password attempt response:", { status: attemptData.status });

    // Check if email verification is needed after password
    if (attemptData.status === "needs_first_factor") {
      logger.info(`Email verification required for ${email}`);

      // Prepare email code verification
      await client.post(
        `/v1/client/sign_ins/${attemptData.id}/prepare_first_factor`,
        { strategy: "email_code" }
      );

      return {
        success: false,
        requires2FA: true,
        signInId: attemptData.id,
        message: "Verification code sent to your email",
      };
    }

    // Check if second factor is needed
    if (attemptData.status === "needs_second_factor") {
      logger.info(`2FA required for ${email}, sending email code...`);
      await client.post(
        `/v1/client/sign_ins/${attemptData.id}/prepare_second_factor`,
        { strategy: "email_code" }
      );

      return {
        success: false,
        requires2FA: true,
        signInId: attemptData.id,
        message: "2FA code sent to your email",
      };
    }

    if (attemptData.status === "complete") {
      logger.info(`Sign-in complete for ${email}`);
      return {
        success: true,
        requires2FA: false,
        sessionToken: attemptData.last_active_session_id,
      };
    }

    logger.warn(`Unexpected sign-in status: ${attemptData.status}`);
    return { success: false, error: `Sign-in incomplete. Please try again.` };
  } catch (error) {
    logger.error("Error in signInWithEmailPassword:", error);
    const errorMessage = error.response?.data?.errors?.[0]?.message || error.message;
    return { success: false, error: errorMessage || "Invalid credentials." };
  }
}

async function verifySignUpCode(signUpId, code) {
  const client = createClerkApiClient();

  try {
    logger.info(`Verifying sign-up code for signUpId: ${signUpId}`);

    const response = await client.post(
      `/v1/client/sign_ups/${signUpId}/attempt_verification`,
      { strategy: "email_code", code }
    );

    const signUpData = response.data.response || response.data;
    logger.info("Sign-up verification response:", { status: signUpData.status });

    if (signUpData.status === "complete") {
      logger.info(`Sign-up complete for signUpId ${signUpId}`);
      return { success: true, sessionToken: signUpData.created_session_id };
    }

    return { success: false, error: "Invalid verification code" };
  } catch (error) {
    logger.error("Error in verifySignUpCode:", error);
    const errorMessage = error.response?.data?.errors?.[0]?.message || error.message;
    return { success: false, error: errorMessage || "Invalid verification code." };
  }
}

async function verify2FACode(signInId, code) {
  const client = createClerkApiClient();

  try {
    logger.info(`Verifying 2FA code for signInId: ${signInId}`);

    // Try first factor verification first
    let response;
    try {
      response = await client.post(
        `/v1/client/sign_ins/${signInId}/attempt_first_factor`,
        { strategy: "email_code", code }
      );
    } catch (firstFactorError) {
      // If first factor fails, try second factor
      logger.info("First factor failed, trying second factor");
      response = await client.post(
        `/v1/client/sign_ins/${signInId}/attempt_second_factor`,
        { strategy: "email_code", code }
      );
    }

    const signInData = response.data.response || response.data;
    logger.info("2FA verification response:", { status: signInData.status });

    if (signInData.status === "complete") {
      logger.info(`2FA complete for signInId ${signInId}`);
      return { success: true, sessionToken: signInData.last_active_session_id };
    }

    return { success: false, error: "Invalid verification code" };
  } catch (error) {
    logger.error("Error in verify2FACode:", error);
    const errorMessage = error.response?.data?.errors?.[0]?.message || error.message;
    return { success: false, error: errorMessage || "Invalid verification code." };
  }
}

module.exports = {
  signUpWithEmailPassword,
  signInWithEmailPassword,
  verify2FACode,
  verifySignUpCode,
};